src/natcap/invest/spec.py

```diff
@@ -0,0 +1,2042 @@
+                return get_message('INVALID_VALUE').format(condition=self.expression)
+
+    @staticmethod
+    def format_column(col, *args):
+        """Format a column of a pandas dataframe that contains NumberInput values.
+
+        Values are cast to float.
+
+        Args:
+            col: Column of a pandas dataframe to format
+
+        Returns:
+            Transformed dataframe column
+        """
+        return col.astype(float)
+
+@dataclasses.dataclass
+class IntegerInput(Input):
+    """An integer input, or parameter, of an invest model."""
+    type: typing.ClassVar[str] = 'integer'
+
+    def validate(self, value):
+        """Validate a value against the requirements for this input.
+
+        Args:
+            value: The value to validate.
+
+        Returns:
+            A string error message if an error was found.  ``None`` otherwise.
+        """
+        try:
+            # must first cast to float, to handle both string and float inputs
+            as_float = float(value)
+            if not as_float.is_integer():
+                return get_message('NOT_AN_INTEGER').format(value=value)
+        except (TypeError, ValueError):
+            return get_message('NOT_A_NUMBER').format(value=value)
+
+    @staticmethod
+    def format_column(col, *args):
+        """Format a column of a pandas dataframe that contains IntegerInput values.
+
+        Values are cast to `pandas.Int64Dtype`.
+
+        Args:
+            col: Column of a pandas dataframe to format
+
+        Returns:
+            Transformed dataframe column
+        """
+        return col.astype(pandas.Int64Dtype())
+
+
+@dataclasses.dataclass
+class RatioInput(Input):
+    """A ratio input, or parameter, of an invest model.
+
+    A ratio is a proportion expressed as a value from 0 to 1 (in contrast to a
+    percent, which ranges from 0 to 100). Values are restricted to the
+    range [0, 1].
+    """
+    type: typing.ClassVar[str] = 'ratio'
+
+    def validate(self, value):
+        """Validate a value against the requirements for this input.
+
+        Args:
+            value: The value to validate.
+
+        Returns:
+            A string error message if an error was found.  ``None`` otherwise.
+        """
+        try:
+            as_float = float(value)
+        except (TypeError, ValueError):
+            return get_message('NOT_A_NUMBER').format(value=value)
+
+        if as_float < 0 or as_float > 1:
+            return get_message('NOT_WITHIN_RANGE').format(
+                value=as_float,
+                range='[0, 1]')
+
+    @staticmethod
+    def format_column(col, *args):
+        """Format a column of a pandas dataframe that contains RatioInput values.
+
+        Values are cast to float.
+
+        Args:
+            col: Column of a pandas dataframe to format
+
+        Returns:
+            Transformed dataframe column
+        """
+        return col.astype(float)
+
+@dataclasses.dataclass
+class PercentInput(Input):
```

Comment:
Perhaps this could inherit from `NumberInput`? Then might not need the `format_column` static method. And `validate` could call `super().validate()`, if desired.

Reply:
Sure, I did not do this originally because in the old model spec, `number`s have attributes that don't apply to `percent` and `ratio`s. But I think it's fine to add those attributes for the sake of a sensible inheritance structure.